service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/azure/common/CloneControlledAzureResourceFlight.java

```diff
@@ -0,0 +1,186 @@
+import bio.terra.workspace.service.resource.controlled.flight.create.GetAzureCloudContextStep;
+import bio.terra.workspace.service.resource.controlled.flight.update.RetrieveControlledResourceMetadataStep;
+import bio.terra.workspace.service.resource.controlled.model.ControlledResource;
+import bio.terra.workspace.service.resource.controlled.model.StepRetryRulePair;
+import bio.terra.workspace.service.resource.model.CloningInstructions;
+import bio.terra.workspace.service.resource.model.WsmResourceType;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public abstract class CloneControlledAzureResourceFlight extends Flight {
+
+  protected CloneControlledAzureResourceFlight(
+      FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    FlightUtils.validateRequiredEntries(
+        inputParameters,
+        WorkspaceFlightMapKeys.ResourceKeys.RESOURCE,
+        JobMapKeys.AUTH_USER_INFO.getKeyName(),
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_RESOURCE_ID,
+        WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID);
+
+    var flightBeanBag = FlightBeanBag.getFromObject(applicationContext);
+
+    var sourceResource =
+        FlightUtils.getRequired(
+            inputParameters,
+            WorkspaceFlightMapKeys.ResourceKeys.RESOURCE,
+            ControlledResource.class);
+    if (!inputParameters.containsKey(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME)) {
+      inputParameters.put(
+          WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_NAME, sourceResource.getName());
+    }
+    var userRequest =
+        inputParameters.get(JobMapKeys.AUTH_USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
+    var destinationWorkspaceId =
+        inputParameters.get(
+            WorkspaceFlightMapKeys.ControlledResourceKeys.DESTINATION_WORKSPACE_ID, UUID.class);
+
+    boolean mergePolicies =
+        Optional.ofNullable(
+                inputParameters.get(WorkspaceFlightMapKeys.MERGE_POLICIES, Boolean.class))
+            .orElse(false);
+    var cloningInstructions =
+        Optional.ofNullable(
+                inputParameters.get(
+                    WorkspaceFlightMapKeys.ResourceKeys.CLONING_INSTRUCTIONS,
+                    CloningInstructions.class))
+            .orElse(sourceResource.getCloningInstructions());
+
+    List<StepRetryRulePair> resourceCloningSteps = new ArrayList<>();
+
+    if (cloningInstructions != CloningInstructions.COPY_NOTHING) {
+      // Get the cloud context and store it in the working map
+      resourceCloningSteps.add(
+          new StepRetryRulePair(
+              new GetAzureCloudContextStep(
+                  destinationWorkspaceId, flightBeanBag.getAzureCloudContextService()),
+              RetryRules.shortDatabase()));
```

Comment:
other than this cloud context, is there any reason this is azure specific?

Reply:
I didn't look much at the other cloning implementations and naively assumed that they differed more across different cloud contexts, though it also makes sense that the cloud context may be the primary differentiator. It also feels like a much larger lift to refactor existing GCP clones, and much higher likelihood I accidentally break something I don't understand.